### ISSUE TITLE
aws-sdk-cpp 1.10.22: undefined reference to `Aws::S3::S3EndpointRules::Rules'

### ISSUE THREAD

**Issue by yumetodo** [msys2/MINGW-packages#14479]
### Description

Failed to link because `Aws::S3::S3EndpointRules::Rules` not found.

Below image of Dependencies result shows that "C:\msys64\mingw32\bin\libaws-cpp-sdk-s3.dll" has no symbol of `Aws::S3::S3EndpointRules::Rules`(sorted by Function name).

![image](https://user-images.githubusercontent.com/10869046/206472087-79ae169c-ae85-48f8-b97e-fa7c4b5403c5.png)

### Verification

- [X] I have verified that my MSYS2 is up-to-date before submitting the report (see https://www.msys2.org/docs/updating/)

### Windows Version

Microsoft Windows 10 [Version 19044.2251]

### MINGW environments affected

- [ ] MINGW64
- [X] MINGW32
- [ ] UCRT64
- [ ] CLANG64
- [ ] CLANG32
- [ ] CLANGARM64

### Expected behavior

no link error

### Actual behavior

```
undefined reference to `Aws::S3::S3EndpointRules::Rules'
```

### Repro steps

`main.cpp`

```cpp
#include <aws/core/Aws.h>
#include <aws/s3/S3Client.h>
#include <aws/core/utils/StringUtils.h>
int main()
{
    auto s3_client = Aws::MakeShared<Aws::S3::S3Client>("S3Client");
    std::cout << std::static_pointer_cast<void>(s3_client).get() << std::endl;
}
```

`CMakeLists.txt`

```cmake
cmake_minimum_required(VERSION 3.24)
enable_language(CXX)
set(CMAKE_CXX_STANDARD 20) # C++20...
set(CMAKE_CXX_STANDARD_REQUIRED ON) #...is required...
set(CMAKE_CXX_EXTENSIONS OFF) #...without compiler extensions like gnu++11
project(aws_link_error_sample)
find_package(AWSSDK REQUIRED COMPONENTS s3)
add_executable(aws_link_error_sample main.cpp)
target_link_libraries(aws_link_error_sample ${AWSSDK_LINK_LIBRARIES})
```

```sh
$cmake -B build_gcc -G "MSYS Makefiles" .
-- The CXX compiler identification is GNU 12.2.0
-- Detecting CXX compiler ABI info
-- Detecting CXX compiler ABI info - done
-- Check for working CXX compiler: C:/msys64/mingw32/bin/c++.exe - skipped
-- Detecting CXX compile features
-- Detecting CXX compile features - done
-- The C compiler identification is GNU 12.2.0
-- Detecting C compiler ABI info
-- Detecting C compiler ABI info - done
-- Check for working C compiler: C:/msys64/mingw32/bin/cc.exe - skipped
-- Detecting C compile features
-- Detecting C compile features - done
-- Found AWS SDK for C++, Version: 1.10.22, Install Root:C:/msys64/mingw32, Platform Prefix:, Platform Dependent Libraries: Userenv;version;ws2_32;Bcrypt;Wininet;winhttp
-- Components specified for AWSSDK: s3, application will be depending on libs: aws-cpp-sdk-s3;aws-cpp-sdk-core
-- Try finding aws-cpp-sdk-core
-- Performing Test CMAKE_HAVE_LIBC_PTHREAD
-- Performing Test CMAKE_HAVE_LIBC_PTHREAD - Success
-- Found Threads: TRUE
-- Found aws-cpp-sdk-core
-- Try finding aws-cpp-sdk-s3
-- Found aws-cpp-sdk-s3
-- Configuring done
-- Generating done
-- Build files have been written to: C:/msys64/home/OPM005562/aws_link_error_sample/build_gcc

$LC_ALL=C cmake --build build_gcc/
[ 50%] Linking CXX executable aws_link_error_sample.exe
C:/msys64/mingw32/bin/../lib/gcc/i686-w64-mingw32/12.2.0/../../../../i686-w64-mingw32/bin/ld.exe: CMakeFiles/aws_link_error_sample.dir/objects.a(main.cpp.obj):main.cpp:(.text$_ZN3Aws2S38Endpoint18S3EndpointProviderC1Ev[__ZN3Aws2S38Endpoint18S3EndpointProviderC1Ev]+0xf): undefined reference to `Aws::S3::S3EndpointRules::Rules'
collect2.exe: error: ld returned 1 exit status
make[2]: *** [CMakeFiles/aws_link_error_sample.dir/build.make:112: aws_link_error_sample.exe] Error 1
make[1]: *** [CMakeFiles/Makefile2:83: CMakeFiles/aws_link_error_sample.dir/all] Error 2
make: *** [Makefile:91: all] Error 2
```

### workaround

Downgrade to 1.9.377, which version doesn't have `Aws::S3::S3EndpointRules` so that work fine.
On mingw32 console with Administrator privileges(to create symbolic link by bsdtar),

```
$pacman -S base_devel git
$git clone --filter=blob:none git@github.com:msys2/MINGW-packages.git
$cd MINGW-packages
$git switch -c tmp/aws_sdk_cpp/1.9 8ce06cffcecc805ac6d247b34cb9def438b34f93
$makepkg -s
$pacman -U mingw-w64-i686-aws-sdk-cpp-1.9.373-1-any.pkg.tar.zst
$cp  /etc/pacman.conf  /etc/pacman.conf.origin
$nano  /etc/pacman.conf
$diff -u /etc/pacman.conf.origin /etc/pacman.conf
--- /etc/pacman.conf.origin     2022-12-09 01:07:20.208532000 +0900
+++ /etc/pacman.conf    2022-12-09 00:59:42.160550800 +0900
@@ -21,7 +21,7 @@
 Architecture = auto

 # Pacman won't upgrade packages listed in IgnorePkg and members of IgnoreGroup
-#IgnorePkg   =
+IgnorePkg   = mingw-w64-i686-aws-sdk-cpp
 #IgnoreGroup =

 #NoUpgrade   =
```

**Comment by yumetodo** [comment]
```
$pacman -Qs aws
local/mingw-w64-i686-aws-c-auth 0.6.21-1
    C99 library implementation of AWS client-side authentication: standard credentials providers and signing (mingw-w64)
local/mingw-w64-i686-aws-c-cal 0.5.17-1
    AWS Crypto Abstraction Layer: Cross-Platform, C99 wrapper for cryptography primitives (mingw-w64).
local/mingw-w64-i686-aws-c-common 0.8.5-1
    Core c99 package for AWS SDK for C. Includes cross-platform primitives, configuration, data structures, and error handling (mingw-w64).
local/mingw-w64-i686-aws-c-compression 0.2.16-1
    C99 implementation of huffman encoding/decoding (mingw-w64)
local/mingw-w64-i686-aws-c-event-stream 0.2.15-1
    C99 implementation of the vnd.amazon.eventstream content-type (mingw-w64)
local/mingw-w64-i686-aws-c-http 0.6.27-1
    C99 implementation of the HTTP/1.1 and HTTP/2 specifications (mingw-w64)
local/mingw-w64-i686-aws-c-io 0.13.11-1
    AWS SDK module to handle all IO and TLS work for application protocols (mingw-w64).
local/mingw-w64-i686-aws-c-mqtt 0.8.0-1
    C99 implementation of the MQTT 3.1.1 specification (mingw-w64)
local/mingw-w64-i686-aws-c-s3 0.2.0-1
    C99 library implementation for communicating with the S3 service, designed for maximizing throughput on high bandwidth EC2 instances (mingw-w64)
local/mingw-w64-i686-aws-c-sdkutils 0.1.7-1
    AWS SDK utility library (mingw-w64)
local/mingw-w64-i686-aws-checksums 0.1.13-1
    Cross-Platform HW accelerated CRC32c and CRC32 with fallback to efficient SW implementations (mingw-w64).
local/mingw-w64-i686-aws-crt-cpp 0.19.1-1
    C++ wrapper around the aws-c-* libraries. Provides Cross-Platform Transport Protocols and SSL/TLS implementations for C++ (mingw-w64)
local/mingw-w64-i686-aws-sdk-cpp 1.10.22-1
    AWS SDK for C++ (mingw-w64)
```

**Comment by yumetodo** [comment]
```sh
$cd /mingw32/include/aws/s3
$rg "S3EndpointRules"
S3EndpointRules.h
15:class S3EndpointRules

S3EndpointProvider.h
14:#include <aws/s3/S3EndpointRules.h>
105:      : S3DefaultEpProviderBase(Aws::S3::S3EndpointRules::Rules)
```

**Comment by yumetodo** [comment]
cc: @MehdiChinoune 
ref: #14448

**Comment by yumetodo** [comment]
I built 0285664ef443b7ca914414fbdf47dc3561a45ffe and cannot found the symbol on dll.

![image](https://user-images.githubusercontent.com/10869046/206485385-1742b0ff-e3ea-4942-b5df-28457790f702.png)

**Comment by yumetodo** [comment]
This class is not exist on aws-sdk-cpp 1.9.x. AFAIK, it was introduced at 1.10.0.

https://sdk.amazonaws.com/cpp/api/1.9.377/namespace_aws_1_1_s3.html

**Comment by yumetodo** [comment]
I built 9341d74cb30dc072efc033372c7d4553c5c95383, a most early 1.10.x package on here, also same.

![image](https://user-images.githubusercontent.com/10869046/206489527-d4a66627-2cd4-44b2-85d1-50fac77459f1.png)

**Comment by yumetodo** [comment]
8ce06cffcecc805ac6d247b34cb9def438b34f93, a latest 1.9.x, success to build sample.

**Comment by Biswa96** [comment]
This seems to be fixed with newer version of aws-sdk-cpp 1.10.36-1 version. Please update all the packages and then try again.
